File: typescript/src/human-review.ts
import {
	type DurableContext,
	withDurableExecution,
} from "@aws/durable-execution-sdk-js";
import { z } from "zod";
import { MODEL } from "./utils/constants.js";
import { converse } from "./utils/converse.js";
import { logger } from "./utils/logger.js";

const ReviewResultSchema = z.object({
	approved: z.boolean(),
	notes: z.string().optional(),
});

/**
 * Human Review Example
 *
 * Demonstrates integrating a human review step in a generative AI workflow.
 * Uses waitForCallback to pause execution until human approval is received.
 */
export const handler = withDurableExecution(
	async (event: { document?: string }, context: DurableContext) => {
		const document = event.document ?? "Sample invoice with amount $1,234.56";

		// Step 1: Extract fields using LLM
		const extractedFields = await context.step("extract fields", async () =>
			converse(
				MODEL.MEDIUM,
				`Extract key fields from this document as JSON: "${document}"`,
			),
		);

		// Step 2: Wait for human review
		const reviewResultStr = await context.waitForCallback<string>(
			"Await Human review",
			async (callbackId) => {
				// In production, send this callbackId to a review system/email
				sendForReview(callbackId, document, extractedFields);
			},
			{ timeout: { days: 7 } },
		);
		const reviewResult = ReviewResultSchema.parse(JSON.parse(reviewResultStr));
		if (!reviewResult.approved) {
			// Step 3: Process based on review outcome
			return { status: "rejected", notes: reviewResult.notes, extractedFields };
		}
		return { status: "approved", extractedFields };
	},
);

const sendForReview = (
	callbackId: string,
	document: string,
	extractedFields: string,
) => {
	// Callback token not logged here - for demo purposes, access it in the console
	logger.info("Review required", { document, extractedFields });
};

File: python/src/human_review.py
import json

from aws_durable_execution_sdk_python import DurableContext, durable_execution
from aws_durable_execution_sdk_python.config import Duration, WaitForCallbackConfig
from pydantic import BaseModel

from src.utils.constants import MODEL
from src.utils.converse import converse


class ReviewResult(BaseModel):
    approved: bool
    notes: str | None = None


def send_for_review(callback_id: str, document: str, extracted_fields: str):
    # Callback token not logged here - for demo purposes, access it in the console
    print(f"Extracted fields: {extracted_fields}")
    print(f"Document: {document}")


@durable_execution
def handler(event: dict, context: DurableContext):
    event = event or {}
    document = event.get("document", "Sample invoice with amount $1,234.56")

    extracted_fields = context.step(
        lambda _: converse(
            MODEL["MEDIUM"],
            f'Extract key fields from this document as JSON: "{document}"',
        ),
        "extract fields",
    )

    review_result_str = context.wait_for_callback(
        lambda callback_id, _: send_for_review(callback_id, document, extracted_fields),
        "Await Human review",
        WaitForCallbackConfig(timeout=Duration.from_days(7)),
    )
    review_result_json = json.loads(review_result_str)
    review_result = ReviewResult(**review_result_json)
    if not review_result.approved:
        return {
            "status": "rejected",
            "notes": review_result.notes,
            "extractedFields": extracted_fields,
        }
    return {"status": "approved", "extractedFields": extracted_fields}
